### DISCUSSION TITLE
How can I set a field to NULL using the Redshift Data API?

### DISCUSSION THREAD

**Discussion by JoseAngel1196** [aws/aws-sdk-go-v2#2619] [Q&A] [answered]
According to the official doc, if you're using sql.Parameter:

- You can't set a value to NULL. The Data API interprets it as the literal string NULL. The following example replaces id with the literal string null. Not the SQL NULL value. ([link](https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html#:~:text=You%20can%27t%20set%20a%20value%20to%20NULL.%20The%20Data%20API%20interprets%20it%20as%20the%20literal%20string%20NULL.%20The%20following%20example%20replaces%20id%20with%20the%20literal%20string%20null.%20Not%20the%20SQL%20NULL%20value.))
- You can't set a zero length value. The Data API SQL statement fails. The following example trys to set id with a zero length value and results in a failure of the SQL statement. ([link](https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html#:~:text=You%20can%27t%20set%20a%20zero%20length%20value.%20The%20Data%20API%20SQL%20statement%20fails.%20The%20following%20example%20trys%20to%20set%20id%20with%20a%20zero%20length%20value%20and%20results%20in%20a%20failure%20of%20the%20SQL%20statement.))

It looks like we can't use either an empty string or a nil value, so what can I do besides using a default value or omitting the field if it is empty? In Go, I saw that we have the this field interface [FieldMemberIsNull](https://github.com/aws/aws-sdk-go-v2/blob/7a2e23fdf325d59bb841f445a701b5d47a09032b/service/redshiftdata/types/types.go#L103), but I'm not sure how to use it. Have anyone done this before?

**Comment by JoseAngel1196** [comment]
Hey @lucix-aws , thoughts?

**Comment by lucix-aws** [comment]
Is that 2nd (link) supposed to be a code snippet? Both of those links go to redshift documentation.

**Reply by JoseAngel1196** [reply]
No, I'm just linking what AWS says just for context

**Reply by JoseAngel1196** [reply]
I'm trying to figure out is how to send a field NULL using the Redshift Data API which looks like we don't have a way to do it but wanted to check in with the team first

**Reply by JoseAngel1196** [reply]
Here's the code:

```
if value == "" {
				paramValue = "null" // Can't use "" or nil

			} else {
				paramValue = fmt.Sprintf("%v", value)
			}
			parameters = append(parameters, types.SqlParameter{
				Name:  aws.String(paramName),
				Value: aws.String(paramValue),
			})
```

**Reply by JoseAngel1196** [reply]
When I send a empty string, the API returns with a ValidationException

**Reply by lucix-aws** [reply, marked as answer]
I haven't used redshift, but based on their documentation there it does seem to be flatly impossible to parameterize a NULL as you've surmised. I think you'd have to put it into the `ExecuteStatementInput.Sql` value directly.

Note that `Field` and its variant `FieldMemberIsNull` is an unrelated union that only appears in operation responses, specifically only in `GetStatementResultOutput.Records`. I assume that's how they model actual null values in records etc.

**Reply by JoseAngel1196** [reply]
Yeah, that’s a total bummer. It’s so weird that they’re not using the FieldMember type in the SqlParameter.
